dig_test.go

```diff
@@ -0,0 +1,360 @@
+		// differently, we go from having an io.Writer satisfied by a
+		// *bytes.Buffer to having an interface{} satisfied by *bytes.Buffer;
+		// the fact that the io.Writer interface was involved is lost forever.)
+		c := New()
+		var w io.Writer = &bytes.Buffer{}
+		require.NoError(t, c.Provide(w), "provide failed")
+		require.Error(t, c.Invoke(func(w io.Writer) {
+		}), "expected relying on provided interface to fail")
+		require.NoError(t, c.Invoke(func(b *bytes.Buffer) {
+			assert.NotNil(t, b, "expected to get concrete type of interface")
+		}), "expected relying on concrete type to succeed")
+
+	})
+
+	t.Run("interface constructor", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() io.Writer {
+			return &bytes.Buffer{}
+		}), "provide failed")
+		require.NoError(t, c.Invoke(func(w io.Writer) {
+			require.NotNil(t, w, "invoke got nil interface")
+		}), "invoke failed")
+	})
+
+	t.Run("multiple-type constructor", func(t *testing.T) {
+		c := New()
+		constructor := func() (*bytes.Buffer, []int, error) {
+			return &bytes.Buffer{}, []int{42}, nil
+		}
+		consumer := func(b *bytes.Buffer, nums []int) {
+			assert.NotNil(t, b, "invoke got nil buffer")
+			assert.Equal(t, 1, len(nums), "invoke got empty slice")
+		}
+		require.NoError(t, c.Provide(constructor), "provide failed")
+		require.NoError(t, c.Invoke(consumer), "invoke failed")
+	})
+
+	t.Run("collections and instances of same type", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() []*bytes.Buffer {
+			return []*bytes.Buffer{{}}
+		}), "providing collection failed")
+		require.NoError(t, c.Provide(func() *bytes.Buffer {
+			return &bytes.Buffer{}
+		}), "providing pointer failed")
+	})
+}
+
+func TestProvideRespectsConstructorErrors(t *testing.T) {
+	t.Run("constructor succeeds", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() (*bytes.Buffer, error) {
+			return &bytes.Buffer{}, nil
+		}), "provide failed")
+		require.NoError(t, c.Invoke(func(b *bytes.Buffer) {
+			require.NotNil(t, b, "invoke got nil buffer")
+		}), "invoke failed")
+	})
+	t.Run("constructor fails", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() (*bytes.Buffer, error) {
+			return nil, errors.New("oh no")
+		}), "provide failed")
+
+		var called bool
+		err := c.Invoke(func(b *bytes.Buffer) { called = true })
+		assert.Contains(t, err.Error(), "oh no", "expected to bubble up constructor error")
+		assert.False(t, called, "shouldn't call invoked function when deps aren't available")
+	})
+}
+
+func TestCantProvideUntypedNil(t *testing.T) {
+	t.Fatalf("Calling Container.Provide(nil) crashes the main testing thread. Failing early.")
+	t.Parallel()
+	c := New()
+	assert.Error(t, c.Provide(nil))
+}
+
+func TestCantProvideErrors(t *testing.T) {
+	t.Parallel()
+	c := New()
+
+	assert.Error(t, c.Provide(func() error { return errors.New("foo") }))
+	// TODO: This is another case where we're losing type information, which
```

Comment:
The type comparison works fine for constructors that explicitly return `error`. It doesn't work if the constructor returns a type that implements the error interface. It definitely doesn't work for instances, since we lose interface information by passing instances to `Provide` as `interface{}` (see https://research.swtch.com/interfaces for details). We *could* drop all types that implement `error`, but that's a larger change.

The intention of these two PRs isn't to fix all bugs in dig, but to get to a place where everyone knows what dig's behavior actually is.